### PR TITLE
Barcode translation

### PR DESCRIPTION
**Pull request by mourisl**
Add the option "--barcode-translate FILE"
The file is a two-column tsv/csv file with the original barcode on the second column and the translated barcode on the first column. 
For 10x multiome, the user can put the scATAC-seq barcode on the second column and the scRNA-seq(GEX) barcode on the first column. The output barcode of Chromap will be in GEX barcode space.
The translation table also handles combinatorial barcoding, such as SHARE-seq, where each barcode is comprised of multiple segments. In the translation table, the second column will be for the each segment and the first column will be the name of that segment. In the end, the output barcode of Chromap will be "-" connected segment names. For example, a cell barcode "ATCACGTTTTGGAGGTGTCTTGGC" will be translated to "A1-D3-H12" in the output automatically.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -22,7 +23,7 @@ dir:
 $(exec): $(objs)
 	$(CXX) $(CXXFLAGS) $(objs) -o $(exec) $(LDFLAGS)
 	
-$(objs_dir)/%.o: $(src_dir)/%.cc
+$(objs_dir)/%.o: $(src_dir)/%.cc $(src_dir)/%.h
```
Comment (haowenz) [inline Makefile:26]: This seems not necessary.

File: src/chromap.cc [hunk 2]
```diff
@@ -2968,12 +2987,19 @@ void Chromap<MappingRecord>::MapSingleEndReads() {
...
+  
+  delete[] mm_history;
+  
+  OutputMappingStatistics();
+  if (!is_bulk_data_) {
+    OutputBarcodeStatistics();
+  }
+
+  index.Destroy();
+  //OutputMappingStatistics(num_reference_sequences, mappings_on_diff_ref_seqs_,
```
Comment (haowenz) [inline src/chromap.cc:3001]: Remove this.

File: src/output_tools.h [hunk 3]
```diff
@@ -34,6 +38,103 @@ bool ReadIdLess(const std::pair<uint32_t, MappingRecord> &a,
   return a.second.read_id_ < b.second.read_id_;
 }
 
+// The class for handling barcode convertion.
```
Comment (haowenz) [inline src/output_tools.h:41]: Can we have a separate header named barcode_translator.h, which holds BarcodeTranslator?

File: src/output_tools.h [hunk 4]
```diff
@@ -173,6 +279,7 @@ class OutputTools {
   FILE *peak_output_file_;
   FILE *barcode_output_file_;
   FILE *matrix_output_file_;
+  BarcodeTranslator barcode_translator;
```
Comment (haowenz) [inline src/output_tools.h:282]: barcode_translator_

File: src/mmcache.hpp [hunk 5]
```diff
@@ -88,6 +88,7 @@ class mm_cache {
       cache[i].finger_print_cnt_sum = 0;
       cache[i].activated = 0;
     }
+    memset(head_mm, 0, sizeof(uint64_t) * HEAD_MM_ARRAY_SIZE);
```
Comment (haowenz) [inline src/mmcache.hpp:91]: Can you explain if this will affect the mapping results?

**Review by haowenz — CHANGES_REQUESTED** [review]
Thanks for adding the feature! Can you also update the README accordingly?

File: src/chromap.cc (hunk 2 above)
Comment (mourisl) [reply src/chromap.cc:3001]: It's in line 2995 to be consistent with the paired-end case.

Comment (haowenz) [reply src/chromap.cc:3001]: Didn't make myself clear. Just remove the lines that have the comment.

File: Makefile (hunk 1 above)
Comment (mourisl) [reply Makefile:26]: If not, the change to the .h file seems will not trigger the recompilation.

File: src/mmcache.hpp (hunk 5 above)
Comment (mourisl) [reply src/mmcache.hpp:91]: This does not. It removes some warnings about initialization.

File: src/chromap.cc (hunk 2 above)
Comment (mourisl) [reply src/chromap.cc]: Done.

File: src/output_tools.h (hunk 3 above)
Comment (mourisl) [reply src/output_tools.h]: Done.

File: src/output_tools.h (hunk 4 above)
Comment (mourisl) [reply src/output_tools.h]: Done.

**Review by haowenz — APPROVED** [review]

**Pull request merged**